Poor Support for Digit Separators

There are two errors, one has been previously reported:
1. Digit Separator Copies Digits
2. Digit Separator Doesn't Work In Fraction

## Steps to Reproduce

In the REPL:

```csharp
> 1_2;             // 122
> 122.3_4;     // Fails, (1,6): error CS1525: Unexpected symbol `_4'
```

### Current & Expected Behavior

There are 2 issues, 1). the digit separator should be valid in the fraction component of a number, and 2). the digit separator should not change the value of the number.

## Platforms

**System Information**

I am running Linux, with the following system information:

```bash
$ lsb_release -a
LSB Version:	:core-4.1-amd64:core-4.1-noarch:cxx-4.1-amd64:cxx-4.1-noarch:desktop-4.1-amd64:desktop-4.1-noarch:languages-4.1-amd64:languages-4.1-noarch:printing-4.1-amd64:printing-4.1-noarch
Distributor ID:	Fedora
Description:	Fedora release 30 (Thirty)
Release:	30
Codename:	Thirty
$ mono --version
Mono JIT compiler version 6.8.0.96 (tarball Wed Jan 15 09:00:57 UTC 2020)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           __thread
	SIGSEGV:       altstack
	Notifications: epoll
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          yes(610)
	Suspend:       hybrid
	GC:            sgen (concurrent by default
```

## Other Related Issues

https://github.com/mono/mono/issues/16648
https://github.com/dotnet/csharplang/issues/55#issuecomment-574901596

--- a/mcs/errors/cs1013-3.cs
+++ b/mcs/errors/cs1013-3.cs
@@ -1,0 +1,7 @@
+// CS1013: Invalid number
+// Line : 6
+
+class X
+{
+    static int i = 1_;
+}

--- a/mcs/errors/cs1013-4.cs
+++ b/mcs/errors/cs1013-4.cs
@@ -1,0 +1,7 @@
+// CS1013: Invalid number
+// Line : 6
+
+class X
+{
+    static double i = 1_.2;
+}

--- a/mcs/errors/cs1013-5.cs
+++ b/mcs/errors/cs1013-5.cs
@@ -1,0 +1,7 @@
+// CS1013: Invalid number
+// Line : 6
+
+class X
+{
+    static int i = 1_e1;
+}

--- a/mcs/errors/cs1013-6.cs
+++ b/mcs/errors/cs1013-6.cs
@@ -1,0 +1,7 @@
+// CS1013: Invalid number
+// Line : 6
+
+class X
+{
+    static float i = 1_f;
+}

--- a/mcs/errors/cs1013-7.cs
+++ b/mcs/errors/cs1013-7.cs
@@ -1,0 +1,7 @@
+// CS1013: Invalid number
+// Line : 6
+
+class X
+{
+    static int i = 0x_1;
+}

--- a/mcs/errors/cs1013-8.cs
+++ b/mcs/errors/cs1013-8.cs
@@ -1,0 +1,7 @@
+// CS1013: Invalid number
+// Line : 6
+
+class X
+{
+    static int i = 0b_1;
+}

--- a/mcs/mcs/cs-tokenizer.cs
+++ b/mcs/mcs/cs-tokenizer.cs
@@ -1926,19 +1926,8 @@ namespace Mono.CSharp
 					}
 				}
 
-			digits:
 				decimal_digits (c);
 				c = peek_char ();
-				if (c == '_') {
-
-					do {
-						get_char ();
-						c = peek_char ();
-					} while (c == '_');
-
-					if (c >= '0' && c <= '9')
-						goto digits;
-				}
 			}
 
 			//TODO: Implement rejection of trailing digit separators
@@ -1992,21 +1981,8 @@ namespace Mono.CSharp
 					number_builder [number_pos++] = '+';
 				}
 
-			digits:
-				bool seen_digits = decimal_digits (c);
+				decimal_digits (c);
 				c = peek_char ();
-
-				if (c == '_' && seen_digits) {
-
-					do {
-						get_char ();
-						c = peek_char ();
-					} while (c == '_');
-
-					if (c >= '0' && c <= '9')
-						goto digits;
-				}
-
 			}
 
 			var type = real_type_suffix (c);

--- a/mcs/mcs/cs-tokenizer.cs
+++ b/mcs/mcs/cs-tokenizer.cs
@@ -1575,7 +1575,8 @@ namespace Mono.CSharp
 		{
 			int d;
 			bool seen_digits = false;
-			
+			bool digit_separator = false;
+
 			if (c != -1){
 				if (number_pos == MaxNumberLength)
 					Error_NumericConstantTooLong ();
@@ -1595,6 +1596,13 @@ namespace Mono.CSharp
 					seen_digits = true;
 					continue;
 				} else if (d == '_') {
+					if (!digit_separator) {
+						if (context.Settings.Version < LanguageVersion.V_7)
+							Report.FeatureIsNotAvailable (context, Location, "digit separators");
+
+						digit_separator = true;
+					}
+
 					get_char();
 					continue;
 				}
@@ -1891,7 +1899,6 @@ namespace Mono.CSharp
 #endif
 			number_pos = 0;
 			var loc = Location;
-			bool digit_separator = false;
 
 			if (!dotLead){
 				if (c == '0'){
@@ -1923,12 +1930,6 @@ namespace Mono.CSharp
 				decimal_digits (c);
 				c = peek_char ();
 				if (c == '_') {
-					if (!digit_separator) {
-						if (context.Settings.Version < LanguageVersion.V_7)
-							Report.FeatureIsNotAvailable (context, Location, "digit separators");
-
-						digit_separator = true;
-					}
 
 					do {
 						get_char ();
@@ -1996,12 +1997,6 @@ namespace Mono.CSharp
 				c = peek_char ();
 
 				if (c == '_' && seen_digits) {
-					if (!digit_separator) {
-						if (context.Settings.Version < LanguageVersion.V_7)
-							Report.FeatureIsNotAvailable (context, Location, "digit separators");
-
-						digit_separator = true;
-					}
 
 					do {
 						get_char ();

--- a/mcs/mcs/cs-tokenizer.cs
+++ b/mcs/mcs/cs-tokenizer.cs
@@ -1576,11 +1576,13 @@ namespace Mono.CSharp
 			int d;
 			bool seen_digits = false;
 			bool digit_separator = false;
+			int prev = c;
+			var loc = Location;
 
-			if (c != -1){
+			if (prev != -1){
 				if (number_pos == MaxNumberLength)
 					Error_NumericConstantTooLong ();
-				number_builder [number_pos++] = (char) c;
+				number_builder [number_pos++] = (char) prev;
 			}
 
 			//
@@ -1592,7 +1594,7 @@ namespace Mono.CSharp
 					if (number_pos == MaxNumberLength)
 						Error_NumericConstantTooLong ();
 					number_builder [number_pos++] = (char)d;
-					get_char ();
+					prev = get_char ();
 					seen_digits = true;
 					continue;
 				} else if (d == '_') {
@@ -1603,13 +1605,23 @@ namespace Mono.CSharp
 						digit_separator = true;
 					}
 
-					get_char();
+					if (prev == '.')
+						break;
+
+					if (prev == 'e' || prev == 'E')
+						Error_InvalidNumber ();
+
+					prev = get_char();
 					continue;
 				}
 
 				break;
 			}
 			
+			if (prev == '_') {
+				Error_InvalidNumber ();
+			}
+
 			return seen_digits;
 		}
 
@@ -1778,9 +1790,8 @@ namespace Mono.CSharp
 			int d;
 			ulong ul;
 			bool digit_separator = false;
-			int prev = 0;
-			
-			get_char ();
+			int prev = get_char ();
+
 			while ((d = peek_char ()) != -1){
 				if (is_hex (d)){
 					number_builder [number_pos++] = (char) d;
@@ -1791,6 +1802,9 @@ namespace Mono.CSharp
 				}
 
 				if (d == '_') {
+					if (prev == 'x' || prev == 'X')
+						Error_InvalidNumber ();
+
 					get_char ();
 
 					if (!digit_separator) {
@@ -1835,10 +1849,9 @@ namespace Mono.CSharp
 			int d;
 			ulong ul = 0;
 			bool digit_separator = false;
-			int prev = -1;
 			int digits = 0;
+			int prev = get_char ();
 
-			get_char ();
 			while ((d = peek_char ()) != -1){
 
 				if (d == '0' || d == '1') {
@@ -1857,6 +1870,9 @@ namespace Mono.CSharp
 				}
 
 				if (d == '_') {
+					if (prev == 'b' || prev == 'B')
+						Error_InvalidNumber ();
+
 					get_char ();
 
 					if (!digit_separator) {
@@ -1929,9 +1945,6 @@ namespace Mono.CSharp
 				decimal_digits (c);
 				c = peek_char ();
 			}
-
-			//TODO: Implement rejection of trailing digit separators
-
 
 			//
 			// We need to handle the case of

--- a/mcs/mcs/cs-tokenizer.cs
+++ b/mcs/mcs/cs-tokenizer.cs
@@ -1594,6 +1594,9 @@ namespace Mono.CSharp
 					get_char ();
 					seen_digits = true;
 					continue;
+				} else if (d == '_') {
+					get_char();
+					continue;
 				}
 
 				break;
@@ -1919,7 +1922,6 @@ namespace Mono.CSharp
 			digits:
 				decimal_digits (c);
 				c = peek_char ();
-
 				if (c == '_') {
 					if (!digit_separator) {
 						if (context.Settings.Version < LanguageVersion.V_7)

--- a/mcs/tests/test-950.cs
+++ b/mcs/tests/test-950.cs
@@ -6,7 +6,7 @@ public class B
 	{
 		int a = 1_0_3;
 		double b = 0__0e+1_1;
-		int c = 0b__1_0;
-		int d = 0x__F_0;
+		int c = 0b0__1_0;
+		int d = 0x0__F_0;
 	}
 }

--- a/mcs/tests/ver-il-net_4_x.xml
+++ b/mcs/tests/ver-il-net_4_x.xml
@@ -53205,7 +53205,7 @@
   <test name="test-950.cs">
     <type name="B">
       <method name="Void Main()" attrs="150">
-        <size>26</size>
+        <size>23</size>
       </method>
       <method name="Void .ctor()" attrs="6278">
         <size>7</size>
@@ -73855,7 +73855,7 @@
   <test name="test-ref-07.cs">
     <type name="TestMain">
       <method name="Void Main()" attrs="150">
-        <size>7</size>
+        <size>6</size>
       </method>
     </type>
     <type name="Test">
@@ -74509,13 +74509,13 @@
   </test>
   <test name="test-tuple-11.cs">
     <type name="Program">
-	  <method name="Int32 Main()" attrs="150">
-	    <size>70</size>
-	  </method>
-	  <method name="Void .ctor()" attrs="6278">
+      <method name="Int32 Main()" attrs="150">
+        <size>70</size>
+      </method>
+      <method name="Void .ctor()" attrs="6278">
         <size>7</size>
-	  </method>
-	</type>
+      </method>
+    </type>
   </test>
   <test name="test-var-01.cs">
     <type name="Test">

